### PR TITLE
[NINJA] Added option to allow scons to determine if it should skip ninja regeneration.

### PR DESCRIPTION
**Pull request by dmoody256**
## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -135,6 +135,8 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
       Note that these are called for every build command run by SCons. It could have considerable
       performance impact if not used carefully.
       to connect to the server during start up.
+    - Ninja: added option to skip ninja regeneration if scons can determine the ninja file does
```
Comment (bdbaddog) [inline CHANGES.txt]: What is the option? How do you enable? (disable)

File: RELEASE.txt [hunk 2]
```diff
@@ -107,6 +107,8 @@ CHANGED/ENHANCED EXISTING FUNCTIONALITY
         require delayed expansion to be enabled which is currently not supported and is
         typically not enabled by default on the host system. The batch files may also require
         environment variables that are not included by default in the msvc environment.
+- Ninja: added option to skip ninja regeneration if scons can determine the ninja file does
```
Comment (bdbaddog) [inline RELEASE.txt]: same as CHANGES.txt comment

File: SCons/Tool/ninja/__init__.py [hunk 3]
```diff
@@ -66,8 +66,12 @@ def ninja_builder(env, target, source):
     print("Generating:", str(target[0]))
 
     generated_build_ninja = target[0].get_abspath()
-    NINJA_STATE.generate()
-
+    try:
+        NINJA_STATE.generate()
+    except Exception:
+        raise SCons.Errors.BuildError(
+            errstr=f"ERROR: an excetion occurred while generating the ninja file:\n{traceback.format_exc()}",
```
Comment (bdbaddog) [inline SCons/Tool/ninja/__init__.py]: Typo: exception

File: test/ninja/ninja_test_sconscripts/sconstruct_ninja_determinism [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import SCons
+import random
+SetOption('experimental','ninja')
+SetOption('skip_ninja_regen', True)
```
Comment (bdbaddog) [inline test/ninja/ninja_test_sconscripts/sconstruct_ninja_determinism:4]: If I comment this line out, the test still passes? Is that expected?

Comment (dmoody256) [reply test/ninja/ninja_test_sconscripts/sconstruct_ninja_determinism:4]: yeah actually this flag is not relevant to the current test. This PR makes ninja files deterministic, but the flag is for controlling skipping the regeneration and restarting of the daemon. The flag is to protect users who may be using Glob or SCons generated files in command lines. Its assumed if the user enables the flag, they know what they are doing.

I will update this test to also make sure the mtime of the ninja file did not change, which would indicate it was never overwritten.

**Comment by dmoody256** [comment]
@bdbaddog fixed the notes, and updated the test

File: test/ninja/ninja_test_sconscripts/sconstruct_ninja_determinism (hunk 4 above)
Comment (bdbaddog) [reply test/ninja/ninja_test_sconscripts/sconstruct_ninja_determinism:4]: o.k. so run this test with master branch code and it should fail then?

Comment (dmoody256) [reply test/ninja/ninja_test_sconscripts/sconstruct_ninja_determinism:4]: yeah I tested that in the latest

**Pull request merged**